js/app/widgets/dynamicLogo.js

```diff
@@ -0,0 +1,308 @@
+// Copyright 2016 Endless Mobile, Inc.
+
+/* exported DynamicLogo */
+
+const EosKnowledgePrivate = imports.gi.EosKnowledgePrivate;
+const Gio = imports.gi.Gio;
+const Gtk = imports.gi.Gtk;
+const GObject = imports.gi.GObject;
+const Rsvg = imports.gi.Rsvg;
+
+const Knowledge = imports.app.knowledge;
+
+const HEIGHT_WIDTH_RATIO = 1.5;
+
+const Width = {
+    MIN: 50,
+    MAX: 400,
+};
+
+const Height = {
+    MIN: 50,
+    MAX: 400,
+};
+
+/**
+ * Class: DynamicLogo
+ *
+ * This widget implements our dynamic logo system. A logo can be composed
+ * of an image, text or both. This widget is considered dynamic because it
+ * can be resized dynamically depending on its allocation size and it can
+ * switch between different layouts depending on the geometry of its
+ * allocation.
+ */
+const DynamicLogo = new Knowledge.Class({
+    Name: 'DynamicLogo',
+    Extends: Gtk.Widget,
+
+    Properties: {
+         /**
+         * Property: image-uri
+         *
+         * A URI to the image displayed in the logo.
+         */
+        'image-uri': GObject.ParamSpec.string('image-uri', 'image-uri', 'image-uri',
+            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT_ONLY, ''),
+         /**
+         * Property: text
+         *
+         * The text displayed in the logo.
+         */
+        'text': GObject.ParamSpec.string('text', 'text', 'text',
+            GObject.ParamFlags.READWRITE, ''),
+         /**
+         * Property: mode
+         *
+         * The mode of the logo, it can be text, image or full.
+         */
+        'mode': GObject.ParamSpec.string('mode', 'mode', 'mode',
+            GObject.ParamFlags.READWRITE, ''),
+    },
+
+    StyleProperties: {
+        /**
+         * Property: sizing
+         *
+         * -EknDynamicLogo-sizing:
+         *  * 'auto'     - size determined by allocated size. Default.
+         *  * 'size-min' - size determined by min-width and min-height css properties.
+         */
+        'sizing': GObject.ParamSpec.string('sizing', 'sizing', 'sizing',
```

Comment:
This is not a blocker, since it's common in various modules; just a note for the future.

Custom style properties will go away in GTK+ 3.90, which means we'll have to find a replacement for this kind of policy change — likely using an actual GObject property, controlled by the YAML/JSON instead of using CSS.

Alternatively, since GTK+ allows using `min-width` and `min-height`, we could implement `max-width` and `max-height`; e.g.:

  - `auto` would be `max-width: fill-content; max-height: fill-content;`
  - `min-size` would be `max-width: min-content; max-height: min-content;`